rdmd.d

```diff
@@ -916,3 +891,71 @@ void yap(size_t line = __LINE__, T...)(auto ref T stuff)
     debug stderr.writeln(line, ": ", stuff);
     else stderr.writeln(stuff);
 }
+
+/**
+Used to wrap filesystem operations that should also be logged with the
+yap function or affected by dryRun. Append the string "IfLive" to the end
+of the function for it to be skipped during a dry run.
+
+These functions allow the filename to be given once so the log statement
+always matches the operation. Using it also guarantees you won't forget to
+include a `yap` alongside any file operation you want to be logged.
+*/
+struct Filesystem
+{
+static:
+    static auto opDispatch(string func, T...)(T args)
+    {
+        static if (func.endsWith("IfLive"))
+        {
+            enum fileFunc = func[0 .. $ - "IfLive".length];
+            enum skipOnDryRun = true;
```

Comment:
This means that the dry-run check will be performed on the basis of the virtuous developer remembering to use the `IfLive` suffix, rather than on the basis of what the actual action is.

In other words, if I call `Filesystem.mkdirRecurse` (without adding the `IfLive` prefix), the filesystem will be modified even in dry run mode.  That's surely not wanted, is it?

The `IfLive` suffix should therefore be dropped in favour of making a decision on the basis of the action concerned.

Reply:
> This means that the dry-run check will be performed on the basis of the virtuous developer remembering to use the IfLive suffix, rather than on the basis of what the actual action is.

> In other words, if I call Filesystem.mkdirRecurse (without adding the IfLive prefix), the filesystem will be modified even in dry run mode. That's surely not wanted, is it?

Yes, actually this is what is wanted.  Whether or not something should be one on a dry run isn't based on the action itself, it's based on what behavior the caller wants in it's particular scenario.

Reply:
> Whether or not something should be one on a dry run isn't based on the action itself, it's based on what behavior the caller wants in it's particular scenario.

But if this PR is anything to go by, there is no existing use-case for _not_ respecting the dry-run behaviour.  So why implement the possibility?

Note also that respecting the dry-run behaviour by default is the _safe_ option because then you are guaranteeing that the filesystem will not be modified.  Forcing a modification should be the non-default, only-on-request option.

With this in mind it's better to make the dry-run behaviour dependent on the filesystem operation (ensuring that ones that modify it are guaranteed to respect the dry-run behaviour) and implement an ability to override that in future only if a use-case ever.

If there's really a strong wish to have non-dry-run behaviour available to the developer, it should be the non-default option that only occurs on clear developer request, rather than the default option.

Reply:
> But if this PR is anything to go by, there is no existing use-case for not respecting the dry-run behaviour. So why implement the possibility?

This is a behavior you don't want to abstract from the caller.  If there's a possibility that the action won't be executed, the caller should definitely be aware of this instead of abstracting it away.

Reply:
> Note also that respecting the dry-run behaviour by default is the safe option because then you are guaranteeing that the filesystem will not be modified.

This actually isn't true, rdmd will still make file system modifications on a dry run.

Reply:
> This actually isn't true, rdmd will still make file system modifications on a dry run.

However, with this diff as a guide, the vast majority of cases are using the `IfLive` constraint.

It would therefore be better for the _exceptions_ to that to be clearly marked and justified, and have the _default_ behaviour be to do nothing (since that's arguably safer).

The exceptions I've noted so far are:

* the `rename` in the case that removal of `fullExe` fails (shouldn't ever be reached in dry-run code so why not keep it using the safe respect-dry-run option?)

* cleanup of `fullExeTemp` if the build fails (shouldn't be necessary in dry-run mode so why worry about the `remove` call taking account of the dry-run constraints?)

* rename of `fullExeTemp` to `fullExe` (only happens if `!dryRun` anyway so why worry about the `rename` call ignoring dry-run constraints?)

* `write` calls for generation of a stub `main` or for use with `--eval` and `--loop`-generated code (presumably necessary, but it's 2 or 3 total calls).

In other words, most of the current use-cases of non-`IfLife` calls are ones that could be replaced with `IfLive` with no change in actual behaviour.  The real exceptions are very limited.

So, given that default behaviour should be _safe_ behaviour, why not make it so that `--dry-run` will _by default_ be respected for all calls to `Filesystem.opDispatch` that would involve modifying the filesystem?

Reply:
> However, with this diff as a guide, the vast majority of cases are using the IfLive constraint.

I don't know where you got your numbers, there's 5 `IfLive` calls, and 27 non-live calls.   I think this false assumption invalidates the rest of this comment.

Reply:
> I don't know where you got your numbers, there's 5 IfLive calls, and 27 non-live calls.

Is that because you are conflating non-`IfLive` calls that modify the filesystem, with non-`IfLive` ones that do not?

By the way, that's another reason I suggest respecting dry-run be the default behaviour.  Then it's _100% clear which calls will modify the filesystem **even in dry-run mode**_.

That's something worthwhile to distinguish.

Reply:
> So, given that default behaviour should be safe behaviour, why not make it so that --dry-run will by default be respected for all calls to Filesystem.opDispatch that would involve modifying the filesystem?

Are you suggesting that we change the extension to "EvenOnDryRun" or something similar and flip the logic?

Reply:
> Are you suggesting that we change the extension to "EvenOnDryRun" or something similar and flip the logic?

Yes, exactly.

Reply:
> Is that because you are conflating non-IfLive calls that modify the filesystem, with non-IfLive 
ones that do not?

There's 5 calls to `Filesystem` that use `IfLive` and 27 that use `Filesytem` that do not.

> By the way, that's another reason I suggest respecting dry-run be the default behaviour. Then it's 100% clear which calls will modify the filesystem even in dry-run mode.

I think the default behavior of calling a function is that it's going to execute no matter what.  Having a function not execute based on some global variable seems like a non-default behavior.  Attaching an extension in the "always execute" case seems unnecesary:
```
writeAlways
writeIfLive
```
Is this what you were thinking?

Reply:
I'd be ok with both variants having an extension, I think it's a bit verbose and unnecessary but would not oppose it.  If @wilzbach or @CyberShadow would prefer this, I'll make the change.  The proposal is to have some sort of extension that indicates a function will always execute it's operation even if we are on a `dryRun`.  So you'd have something like
```
mkdirIfLive
mkdirEvenOnDryRun
```

Wait, now that I think about it, these name extensions don't make sense for functions that return something, i.e.
```
existsIfLive
existsEvenOnDryRun
```

Reply:
> There's 5 calls to Filesystem that use IfLive and 27 that use Filesytem that do not.

How many of those 27 calls involve invoking `std.file` functionality that modifies the filesystem?

Reply:
> Having a function not execute based on some global variable seems like a non-default behavior.

Picture this from the point of view of someone trying to understand what `rdmd.d` is doing, who isn't already familiar with the codebase.  There are _very_ few cases where in `--dry-run` mode the file system actually gets modified.  So for that reason, it's very useful to have those cases clearly highlighted and visible in the code.

If `--dry-run` mode is by default respected for anything that might modify the filesystem, and exceptions to that behaviour have to be clearly and explicitly requested, then:

* it's harder for a new developer to introduce new code that inadvertently makes a FS modification that wasn't wanted

* it's harder for reviewers to miss the introduction of new code that modifies the FS where it's not wanted

Where is the downside?

Reply:
> There are very few cases where in --dry-run mode the file system actually gets modified. So for that reason, it's very useful to have those cases clearly highlighted and visible in the code.

You're conflating `std.file` with operations that modify the file system.  Most of the function don't modify the file system, they query it or do something related to files.

Reply:
> You're conflating std.file with operations that modify the file system.

Not at all.  Most `std.file` operations indeed do not modify the filesystem and therefore are dry-run-safe.  A few of them do, and therefore it's useful for the _default_ `Filesystem` behaviour to be to make sure that they do not do anything in dry-run mode.  That allows the user to have trust that by default, _all_ calls to anything via `Filesystem` _will not modify the filesystem in dry-run-mode_.

And then, there are the _very few_ cases where we want to override that behaviour and _really_ make a FS modification even in dry-run mode.  If those are on request, then it is VERY clear what they are, and that makes for a much easier-to-understand (and safer) codebase.

Reply:
What's you're proposed solution?

Reply:
Your existing solution goes essentially:

```D
static if (func.endsWith("IfLive"))
{
    enum skipOnDryRun = false;
}

...

static if (skipOnDryRUn)
{
    if (dryRun) return;
}
```

Instead:

```D
static if (func.endsWith("EvenOnDryRun"))
{
    enum forceOnDryRun = true;
}

// explicitly whitelist the functions you plan on allowing to run by default with dry-run
static if (fileFunc.among("exists", "timeLastModified", "isFile", "isDir", "existsAsFile"))
    enum useByDefaultWithDryRun = true;
else
    enum useByDefaultWithDryRun = false;

...

// `forceEvenOnDryRun` overrides the `!useByDefaultWithDryRun`
static if (!useByDefaultWithDryRun && !forceEvenOnDryRun)
{
    if (dryRun) return;
}
```

Reply:
Would we need `EvenOnDryRun` with what `rdmd` does currently?

Reply:
So he's proposing that some functions always either use `IfLive` or `EvenOnDryRun` extensions, and others don't use them at all.

```
Filesystem.mkdirRecurseIfLive
Filesystem.mkdirRecurseEvenOnDryRun

// VS

Filesystem.exists
Filesystem.timeLastModified
```

I would oppose this, but I don't think it's necessary.  To me, `EvenOnDryRun` doesn't really add any information, I think `mkdirRecurse` by itself is already clear that it's going to happen regardless of dry run, but that's just my opinion.  If @CyberShadow or @wilzbach prefer to use the `EvenOnDryRun` postfix, I'll implement it.

Reply:
No, I think the idea is to remove `IfLive` completely and make that "the default".

Reply:
> Would we need `EvenOnDryRun` with what `rdmd` does currently?

There are three cases I identified:

* the `write` call inside `makeEvalFile`, where `--eval` or `--loop` options are being used to auto-generate a D file to be built (and presumably, any cleanup of the same);

* the `write` call inside the `if (addStubMain)` block

There's also a block:

```D
// DMD uses Windows-style command-line parsing in response files
// regardless of the operating system it's running on.
Filesystem.write(rspName, array(map!escapeWindowsArgument(todo)).join(" "));
```

Those are the 3 cases that I found where even-in-dry-run-mode seemed necessary.  Others in the existing PR changes do not seem to have that requirement.

Reply:
> No, I think the idea is to remove the `IfLive` suffix and instead make that "the default", with `EvenOnDryRun` being the opt-out (of the `dryRun` check).

Exactly.  Thanks for clarifying!

Reply:
> No, I think the idea is to remove the IfLive suffix and instead make that "the default", with EvenOnDryRun being the opt-out (of the dryRun check).

Ugh, this sounds like a nightmare.  Some functions would be affected by dryRun and some wouldn't, the only way you would know is if you went and looked at the implementation of `Filesystem` and followed the logic.

Reply:
The first two sound like bugs where `rdmd` is not honoring `--dry-run`, even when it should.

The third one sort of does too, but I don't know how much `rdmd` can keep meaningfully going forward after pretending to generate the dependencies.

Reply:
>  Some functions would be affected by `dryRun` and some wouldn't, the only way you would know is if you went and looked at the implementation of `Filesystem` and followed the logic.

That's already the case, is it not?

Reply:
No, the current PR always executes the function, unless you specify `IfLive`.

Reply:
> Ugh, this sounds like a nightmare. Some functions would be affected by dryRun and some wouldn't, 

Well, if we consistently implement this according to the obvious rule ("does it change something on the filesystem? then it should be affected by --dry-run"), then that shouldn't be an issue... the proposal is enticing but I agree the current design is simpler.

Maybe we should just merge this and call it a day. I think we're out of low-hanging fruit, and it seems like we're losing the forest for the trees with all this debate over trivia. We should try to improve the status quo (in a way that preferably doesn't compromise future improvements), and move on to the next problem.

Reply:
> No, the current PR always executes the function, unless you specify `IfLive`.

It's usually pretty obvious from the function name which of them change things (`touch`, `rename`, `remove`, `write`, `mkdir`, ...) and which of them do not (`exists`, `isFile`, `isDir`, ...).